Mitschriften System-Entwicklung 17.05.2021 (#47)
importPI19fromDHGE/dhge-pi19-sem4: SWE-KASCHE/README.md

**MPI**

- Prozess auf einer anderen Maschine starten
- Daten für die Ausführung (und Ergebnis) über standardisiertes Protokoll zwischen Maschinen übertragen

### Architektur-Modelle der Ablauf-Sicht

> **Zielperson:** Integrator 

Author: ```suggestion
**Zielperson:** Integrator
```

    - Call and Return
      - Hauptprogramm, Unterprogramm 1, 2, 3
    - Master Slave
      - Manager, GUI, Sensor, Aktor
  - Ereignis Steuerung
    - Selective Broadcast
      - Event-Handler, Subsystem 1, 2, 3
      - Events werden von einem System geraised und vom Event-Handler behandelt
      - Events werden nicht an alle Systeme, sondern immer nur an bestimmte gesendet (selective)
      - externe Events können auch an den Event Handler gereicht werden
      - Bsp.: GUI, Microservices
      - Nachteil: in der gezeigten Ausbaustufe: keine Info über Subsystem - Erfolg <!-- ??? -->, Scheduling schierig, zentrale Stelle des Event-Handlers beim Ausfall
      - Vorteil: zentrale Stelle für Vorverarbeitung, Konfiguration, Priorisierung
    - Interrupt
      - Interrupt-Dispatcher, Handler 1, Prozess 1, Handler 2, Prozess 2
      - Bspw.: Keyboard-Handler (Lesen des ASCII-Codes, Prozess starten + Keycode Info mitgeben, Info an Dispatcher)

### Architektur-Muster der Logischen Sicht

**ZielGruppe:** Benutzer
- Verfeinerung des Analyse-Modells
- Blickpunkte des Benutzers (z. Bsp. durch Aktivitätsdiagramm)
- Schwarzer Punkte: Start, Punkte mit Kreis: Stop
- ggf. Swimlanes verwenden

> ## Einflussfaktoren

Author: ```suggestion
## Einflussfaktoren

```

- Einsatzbedingungen
  - Einsetzbarkeit (Skalierbarkeit)
  - Zielplattform (Integration wird notwendig)
  - vorhandene Schnittstellen
  - Netzverteilung
  - Benutzerzielgruppe / Service-Modell
  - Mandantenfähigkeit
  - Hilfesystem
  - Entwicklungsmodell beim System-Entwurf
  - nicht-funktionale Anforderungen

>   

Author: ```suggestion

```

## Grundsatzentscheidungen
- welche zusätzlichen Dienstleistungen werde ich anbieten?
- welche Dienstleistungen können wegfallen?
- Systemumgebung
- Ausbaustufen gewünscht
- Hilfesystem
  - schmale Schnittstelle zum Hilfesystem (XML, JSON, CSV, PDF, MD, ...)

> - Verteilung im Netz (Client-Server, Web-Architektur) 

Author: ```suggestion
- Verteilung im Netz (Client-Server, Web-Architektur)
```

**MPI**

- Prozess auf einer anderen Maschine starten
- Daten für die Ausführung (und Ergebnis) über standardisiertes Protokoll zwischen Maschinen übertragen

### Architektur-Modelle der Ablauf-Sicht

**Zielperson:** Integrator

> - Definition von Nebenläufigen System-Einheiten (Prozesse, Threads etc.)

Reviewer: ```suggestion
- Definition von nebenläufigen System-Einheiten (Prozesse, Threads etc.)
```

- Steuerung der Abfolge von Einzelfunktionen
- Synchronisation / Koordination
- Reaktion auf Ereignisse
- Darstellung: Sequenz-Diagramme
  > Stichwort: Swimlanes

- Muster:
  - Zentrale Steuerung
    - Call and Return

>       - Hauptprogramm, Unterprogramm 1, 2, 3

Reviewer: ```
      - Hauptprogramm, Unterprogramm 1, 2, 3
```

Oh, damit wüsste ich nächste Woche schon nix mehr anzufangen :D

Author: Same tbh, für den Abschnitt würden sich ein paar Zeichnungen anbieten

Reviewer 2: wollen wir quasi ein ToDo-Issue dafür aufmachen?

    - Master Slave
      - Manager, GUI, Sensor, Aktor

>   - Ereignis Steuerung

Reviewer: ```suggestion
  - Ereignis-Steuerung
```

    - Selective Broadcast
      - Event-Handler, Subsystem 1, 2, 3
      - Events werden von einem System geraised und vom Event-Handler behandelt
      - Events werden nicht an alle Systeme, sondern immer nur an bestimmte gesendet (selective)
      - externe Events können auch an den Event Handler gereicht werden
      - Bsp.: GUI, Microservices

>       - Nachteil: in der gezeigten Ausbaustufe: keine Info über Subsystem - Erfolg <!-- ??? -->, Scheduling schierig, zentrale Stelle des Event-Handlers beim Ausfall

Reviewer: 😂

>       - Nachteil: in der gezeigten Ausbaustufe: keine Info über Subsystem - Erfolg <!-- ??? -->, Scheduling schierig, zentrale Stelle des Event-Handlers beim Ausfall

Reviewer: ```suggestion
      - Nachteil: in der gezeigten Ausbaustufe: keine Info über Subsystem - Erfolg <!-- ??? -->, Scheduling schwierig, Event-Handler ist Single Point Of Failure
```

      - Vorteil: zentrale Stelle für Vorverarbeitung, Konfiguration, Priorisierung
    - Interrupt
      - Interrupt-Dispatcher, Handler 1, Prozess 1, Handler 2, Prozess 2
      - Bspw.: Keyboard-Handler (Lesen des ASCII-Codes, Prozess starten + Keycode Info mitgeben, Info an Dispatcher)

### Architektur-Muster der Logischen Sicht

> **ZielGruppe:** Benutzer

Reviewer: ```suggestion
**Zielgruppe:** Benutzer
```
[Beweis](https://www.duden.de/rechtschreibung/Zielgruppe)

    - Call and Return
      - Hauptprogramm, Unterprogramm 1, 2, 3
    - Master Slave
      - Manager, GUI, Sensor, Aktor
  - Ereignis-Steuerung
    - Selective Broadcast
      - Event-Handler, Subsystem 1, 2, 3
      - Events werden von einem System geraised und vom Event-Handler behandelt
      - Events werden nicht an alle Systeme, sondern immer nur an bestimmte gesendet (selective)
      - externe Events können auch an den Event Handler gereicht werden
      - Bsp.: GUI, Microservices
      - Nachteil: in der gezeigten Ausbaustufe: keine Info über Subsystem - Erfolg <!-- ??? -->, Scheduling schwierig, Event-Handler ist Single Point Of Failure
      - Vorteil: zentrale Stelle für Vorverarbeitung, Konfiguration, Priorisierung
      - ![Selektive Broadcast](assets/selective_broadcast.jpg)<!--width=600px-->
    - Interrupt
      - Interrupt-Dispatcher, Handler 1, Prozess 1, Handler 2, Prozess 2
      - Bspw.: Keyboard-Handler (Lesen des ASCII-Codes, Prozess starten + Keycode Info mitgeben, Info an Dispatcher)

### Architektur-Muster der Logischen Sicht

**Zielgruppe:** Benutzer

- Verfeinerung des Analyse-Modells
- Blickpunkte des Benutzers (z. Bsp. durch Aktivitätsdiagramm)

> - Schwarzer Punkte: Start, Punkte mit Kreis: Stop

Reviewer 2: ```suggestion
- Schwarze Punkte: Start, Punkte mit Kreis: Stop
```

- ggf. Swimlanes verwenden

## Einflussfaktoren

- Einsatzbedingungen
  - Einsetzbarkeit (Skalierbarkeit)
  - Zielplattform (Integration wird notwendig)
  - vorhandene Schnittstellen
  - Netzverteilung
  - Benutzerzielgruppe / Service-Modell
  - Mandantenfähigkeit
  - Hilfesystem
  - Entwicklungsmodell beim System-Entwurf
  - nicht-funktionale Anforderungen

## Grundsatzentscheidungen

> - welche zusätzlichen Dienstleistungen werde ich anbieten?

Reviewer 2: ```suggestion
- Welche zusätzlichen Dienstleistungen werde ich anbieten?
```

> - welche Dienstleistungen können wegfallen?

Reviewer 2: ```suggestion
- Welche Dienstleistungen können wegfallen?
```
